controller/src/main/java/io/pravega/controller/store/stream/PravegaTablesStream.java

```diff
@@ -278,15 +277,32 @@ private String getWritersTableName(String id) {
     }
 
     @Override
-    public CompletableFuture<Void> createSubscriber(String newSubscriber) {
+    public CompletableFuture<Void> createSubscriber(String newSubscriber, long operationGeneration) {
         final StreamSubscriber newSubscriberRecord = new StreamSubscriber(newSubscriber, ImmutableMap.of(), System.currentTimeMillis());
         return getMetadataTable()
                 .thenCompose(metadataTable -> getSubscriberSetRecord(true)
-                        .thenCompose(subscriberSetRecord -> storeHelper.updateEntry(metadataTable, SUBSCRIBER_SET_KEY,
-                                SubscriberSet.add(subscriberSetRecord.getObject(), newSubscriber).toBytes(), subscriberSetRecord.getVersion())
-                                .thenCompose(v -> storeHelper.addNewEntryIfAbsent(metadataTable, getKeyForSubscriber(newSubscriber), newSubscriberRecord.toBytes()))
-                                .thenAccept(v -> storeHelper.invalidateCache(metadataTable, SUBSCRIBER_SET_KEY))
-                                .thenAccept(v -> storeHelper.invalidateCache(metadataTable, getKeyForSubscriber(newSubscriber)))));
+                   .thenCompose(subscriberSetRecord -> {
+                       if (subscriberSetRecord.getObject().getSubscribers().containsKey(newSubscriber)) {
+                           // update Subscriber generation
+                           Long generation = subscriberSetRecord.getObject().getSubscribers().get(newSubscriber);
+                           if (generation.longValue() < operationGeneration) {
+                               storeHelper.updateEntry(metadataTable, SUBSCRIBER_SET_KEY,
+                                SubscriberSet.update(subscriberSetRecord.getObject(), newSubscriber, generation).toBytes(),
+                                                                         subscriberSetRecord.getVersion())
+                                .thenAccept(v -> storeHelper.invalidateCache(metadataTable, SUBSCRIBER_SET_KEY));
+                           }
+                      } else {
+                           // add new Subscriber
+                           storeHelper.updateEntry(metadataTable, SUBSCRIBER_SET_KEY,
+                                   SubscriberSet.add(subscriberSetRecord.getObject(),
+                                           newSubscriber, operationGeneration).toBytes(), subscriberSetRecord.getVersion())
+                                   .thenCompose(v -> storeHelper.addNewEntryIfAbsent(metadataTable,
+                                           getKeyForSubscriber(newSubscriber), newSubscriberRecord.toBytes()))
+                                   .thenAccept(v -> storeHelper.invalidateCache(metadataTable, SUBSCRIBER_SET_KEY))
+                                   .thenAccept(v -> storeHelper.invalidateCache(metadataTable, getKeyForSubscriber(newSubscriber)));
+                       }
+                       return CompletableFuture.completedFuture(null);
```

Comment:
whatever computation you are doing above is done asynchronously and then you are returning a completed future which is not linked to the above computation.. this is incorrect. this method will return with a completed future and caller will assume the work is done while the work would still be happening asynchronously and could fail.. the response future should be linked to the actual processing.

Reply:
ok. fixed.